compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1875,6 +1879,48 @@ impl<'tcx> TyCtxt<'tcx> {
         }
     }
 
+    /// Returns layout of a `async_drop_in_place::{closure}` coroutine
+    ///   (returned from `async fn async_drop_in_place<T>(..)`).
+    /// Layout might be unavailable if the coroutine is tainted by errors.
+    pub fn async_drop_coroutine_layout(self, ty: Ty<'tcx>) -> Option<&'tcx CoroutineLayout<'tcx>> {
+        self.lang_items().async_drop_in_place_poll_fn().and_then(|def_id| {
+            self.mir_shims(InstanceKind::AsyncDropGlue(def_id, ty)).coroutine_layout_raw()
+        })
+    }
+
+    /// Returns layout of a templated (or not) coroutine. Layout might be unavailable if the
+    /// coroutine is tainted by errors.
+    pub fn coroutine_layout(
+        self,
+        def_id: DefId,
+        args: GenericArgsRef<'tcx>,
+    ) -> Option<&'tcx CoroutineLayout<'tcx>> {
+        if Some(def_id) == self.lang_items().async_drop_in_place_poll_fn() {
```

Comment:
since you're only checking that it is this id here, you could also check that the id's parent is the actual function instead of the coroutine, avoiding the new lang item

Reply:
But which def_id should I use for AsyncDropGlue shim?

For async function `async_drop_in_place` there are shim `AsyncDropGlueCtorShim` (for `async_drop_in_place` constructor function, returning future with async drop implementation), and shim `AsyncDropGlue` (for `async_drop_in_place::{closure}` poll function).

If I use constructor function def_id both for constructor and poll shim, will it be ok for instances resolving etc?

Reply:
You'd still use the one of the coroutine. You do have it available here. All I'm saying is to change this if condition to check the parent's language item instead

Reply:
LangItem for `async_drop_in_place_poll_fn` removed, `ctx.parent(cor_def_id) == async_drop_in_place_fn` is now used instead.